src/renderers/shared/fiber/ReactFiberReconciler.js

```diff
@@ -269,15 +272,20 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
 
     getPublicRootInstance(
       container: OpaqueRoot,
-    ): ReactComponent<any, any, any> | I | TI | null {
+    ): ReactComponent<any, any, any> | PI | null {
       const containerFiber = container.current;
       if (!containerFiber.child) {
         return null;
       }
-      return containerFiber.child.stateNode;
+      switch (containerFiber.child.tag) {
+        case HostComponent:
+          return getPublicInstance(containerFiber.child.stateNode);
+        default:
```

Comment:
This case should match on `ClassComponent` only. It won't type check if we use proper disjoint unions on Fibers.

That type error would be correct since this can be other types such as fragments, coroutines etc. Those should do something reasonable like throw, drill down (like findDOMNode does) or return null.